modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py

```diff
@@ -1976,27 +2084,71 @@ def _execute_arrow(self):
         pyarrow.Table
             The resulting table.
         """
-        if isinstance(self._op, FrameNode):
-            if self._partitions.size == 0:
-                return pyarrow.Table.from_pandas(pd.DataFrame({}))
+        result = None
+        stack = [self]
+
+        while stack:
+            frame = stack.pop()
+
+            if callable(frame):
+                frame()
+            elif isinstance(frame._op, FrameNode):
+                if frame._partitions.size == 0:
+                    result = pyarrow.Table.from_pandas(pd.DataFrame({}))
+                else:
+                    assert frame._partitions.size == 1
+                    result = frame._partitions[0][0].get()
+            elif isinstance(frame._op, MaskNode):
+
+                def slice(positions=frame._op.row_positions):
+                    nonlocal result
+                    result = self._arrow_row_slice(result, positions)
+
+                stack.append(slice)
+                stack.append(frame._op.input[0])
+            elif isinstance(frame._op, TransformNode):
+
+                def select(exprs=frame._op.exprs):
+                    nonlocal result
+                    result = self._arrow_select(result, exprs)
+
+                stack.append(select)
+                stack.append(frame._op.input[0])
+            elif isinstance(frame._op, UnionNode):
+
+                def union(op=frame._op, tables={}, input=iter(frame._op.input)):
+                    nonlocal result
+
+                    if (i := next(input, None)) is None:
```

Comment:
can you explain this, when an input could be None? Doesn't the input is the frames to concatenate so they can't be None?

Reply:
This function is called multiple times, for each frame from the input. After processing the last frame, we will get None here.